File: backend/src/orchestrators/market.orchestrator.ts
import { Injectable } from '@nestjs/common';
import { MarketResearchDto } from '../dto/market-research.dto';
import { ResearchAgent } from '../agents/research.agent';
import { CompetitorAgent } from '../agents/competitor.agent';
import { SummaryAgent } from '../agents/summary.agent';

@Injectable()
export class MarketOrchestrator {
  constructor(
    private readonly researchAgent: ResearchAgent,
    private readonly competitorAgent: CompetitorAgent,
    private readonly summaryAgent: SummaryAgent,
  ) {}

  async execute(input: MarketResearchDto) {
    const research = await this.researchAgent.analyze(input);
    const competitor = await this.competitorAgent.analyze(input);
    const summary = await this.summaryAgent.analyze(input);

    return {
      research,
      competitor,
      summary,
    };
  }
}

File: ai-agents/orchestrators/market_orchestrator.py
from agents.competitor_agent import competitor_agent
from agents.research_agent import research_agent
from agents.summary_agent import summary_agent
from schemas import AnalyzeRequest, AnalyzeResponse
from services.openai_client import logger


def run_market_analysis(payload: AnalyzeRequest) -> AnalyzeResponse:
    logger.info(
        "[MarketOrchestrator] start | topic=%s | region=%s | markets=%s",
        payload.topic,
        payload.region,
        ", ".join(payload.markets),
    )

    research = research_agent(payload)
    competitor = competitor_agent(payload)
    summary = summary_agent(payload, research, competitor)

    logger.info("[MarketOrchestrator] assembled final response")

    return AnalyzeResponse(
        topic=payload.topic,
        region=payload.region,
        markets=payload.markets,
        keyMarkets=research.keyMarkets,
        marketInsights=research.marketInsights,
        recentDevelopments=competitor.recentDevelopments,
        externalSignals=competitor.externalSignals,
        overallInsight=summary.overallInsight,
        opportunities=summary.opportunities,
        risks=summary.risks,
    )